File: windows/YtRec.Core/Timecode.cs
using System.Globalization;

namespace YtRec.Core;

/// <summary>Timecode parse/format for "grab a VOD section" (pure logic). Mirrors mac Timecode.swift.</summary>
public static class Timecode
{
    /// <summary>Parse "90" / "5:30" / "1:05:30" / "1:30.5" → seconds; null if invalid.</summary>
    public static double? Parse(string raw)
    {
        var s = raw.Trim();
        if (s.Length == 0) return null;
        var parts = s.Split(':'); // empty subsequences kept: "5:" -> ["5",""]
        if (parts.Length is < 1 or > 3) return null;

        var values = new List<double>(parts.Length);
        foreach (var p in parts)
        {
            if (!double.TryParse(p, NumberStyles.Float, CultureInfo.InvariantCulture, out var v) || v < 0)
                return null;
            values.Add(v);
        }
        return values.Count switch
        {
            1 => values[0],
            2 => values[0] * 60 + values[1],
            _ => values[0] * 3600 + values[1] * 60 + values[2],
        };
    }

    /// <summary>Seconds → "HH:MM:SS" (for yt-dlp).</summary>
    public static string Format(double seconds)
    {
        var total = Math.Max(0, (int)Math.Round(seconds, MidpointRounding.AwayFromZero));
        return $"{total / 3600:D2}:{total % 3600 / 60:D2}:{total % 60:D2}";
    }

    /// <summary>Seconds → short display ("M:SS", or "H:MM:SS" past an hour).</summary>
    public static string FormatShort(double seconds)
    {
        var total = Math.Max(0, (int)Math.Round(seconds, MidpointRounding.AwayFromZero));
        int h = total / 3600, m = total % 3600 / 60, sec = total % 60;
        return h > 0 ? $"{h}:{m:D2}:{sec:D2}" : $"{m}:{sec:D2}";
    }

    /// <summary>yt-dlp --download-sections "*START-END"; requires 0 ≤ start &lt; end, else null.</summary>
    public static string? DownloadSectionArg(double startSec, double endSec)
    {
        if (startSec < 0 || endSec <= startSec) return null;
        return $"*{Format(startSec)}-{Format(endSec)}";
    }

    /// <summary>From two raw strings → (arg, human label), both must be valid and start &lt; end.</summary>
    public static (string Arg, string Label)? Section(string startRaw, string endRaw)
    {
        var a = Parse(startRaw);
        var b = Parse(endRaw);
        if (a is null || b is null) return null;
        var arg = DownloadSectionArg(a.Value, b.Value);
        if (arg is null) return null;
        return (arg, $"{FormatShort(a.Value)}–{FormatShort(b.Value)}");
    }
}

File: mac/Sources/YTRec/Support/Timecode.swift
import Foundation

/// 時間碼解析與格式化（純邏輯，可測試）。用於「只抓 VOD 的某一段」。
enum Timecode {

    /// 解析使用者輸入的時間碼 → 秒。支援「90」「5:30」「1:05:30」「1:30.5」等；無效回 nil。
    static func parse(_ raw: String) -> Double? {
        let s = raw.trimmingCharacters(in: .whitespaces)
        guard !s.isEmpty else { return nil }
        let parts = s.split(separator: ":", omittingEmptySubsequences: false).map(String.init)
        guard (1...3).contains(parts.count) else { return nil }
        var values: [Double] = []
        for p in parts {
            guard let v = Double(p), v >= 0 else { return nil }
            values.append(v)
        }
        switch values.count {
        case 1: return values[0]                                   // 秒
        case 2: return values[0] * 60 + values[1]                  // MM:SS
        default: return values[0] * 3600 + values[1] * 60 + values[2]  // HH:MM:SS
        }
    }

    /// 秒 → "HH:MM:SS"（給 yt-dlp 用）
    static func format(_ seconds: Double) -> String {
        let total = max(0, Int(seconds.rounded()))
        return String(format: "%02d:%02d:%02d", total / 3600, (total % 3600) / 60, total % 60)
    }

    /// 秒 → 顯示用短格式（< 1 小時顯示 M:SS）
    static func formatShort(_ seconds: Double) -> String {
        let total = max(0, Int(seconds.rounded()))
        let h = total / 3600, m = (total % 3600) / 60, s = total % 60
        return h > 0 ? String(format: "%d:%02d:%02d", h, m, s) : String(format: "%d:%02d", m, s)
    }

    /// yt-dlp `--download-sections` 參數 "*START-END"；需 0 ≤ start < end，否則回 nil。
    static func downloadSectionArg(startSec: Double, endSec: Double) -> String? {
        guard startSec >= 0, endSec > startSec else { return nil }
        return "*\(format(startSec))-\(format(endSec))"
    }

    /// 從兩個使用者輸入字串直接算出 section 參數＋人話標籤（兩者都要有效且 start < end）。
    static func section(from startRaw: String, to endRaw: String) -> (arg: String, label: String)? {
        guard let a = parse(startRaw), let b = parse(endRaw),
              let arg = downloadSectionArg(startSec: a, endSec: b) else { return nil }
        return (arg, "\(formatShort(a))–\(formatShort(b))")
    }
}
